### DISCUSSION TITLE
docs: Describe how to use UseRequestContext for Queues preventing race conditions

### DISCUSSION THREAD

**Discussion by rstoermer** [mikro-orm/mikro-orm#3683] [Q&A]
Hey,

I am using Mikro ORM and like it very much. I got a small glitch in my application however which is probably not a bug, but a usage failure. I am however not sure how to fix this and can't find anything in the documentation: https://mikro-orm.io/docs/identity-map /  https://mikro-orm.io/docs/usage-with-nestjs/#request-scoped-handlers-in-queues.

In my Application I am receiving Messages via a queue with the following code (Simplified):

```typescript
@UseRequestContext()
  async handleIncomingMessage(
    msg:
      | ResourceUpdatedRequestDTO<TrackedResources>
      | ResourceDeletedRequestDTO,
    amqpMsg: Message,
  ) {
    try {
      const { resourceType } = msg;

      // If we don't track a resource yet, we return
      if (!this.resourceService.trackableResources.includes(resourceType)) {
        return;
      }

      this.logger.log(
        `Received message with ${amqpMsg.fields.routingKey} for resource ${resourceType}`,
      );

      await this.resourceService.resourceUpdatedHandler(
        resourceType,
        (msg as ResourceDeletedRequestDTO).resourceId,
        new Date(amqpMsg.properties.timestamp),
        amqpMsg.properties.correlationId,
      );
      return;
    } catch (error: any) {
      this.logger.error(
        `${error.message} while processing ${JSON.stringify(msg)}, : ${
          error.stack
        }`,
      );
    }
  }
````
Basically I am creating a RequestContext for every received message. However, what happens is, if I received a message for the same Resource at almost the same time, I get a race condition, where from the first message a Resource is created, but the second message creates a duplicate key error.

This basically leads to such a log message:

[resource-history] {"level":30,"time":"2022-11-03T08:36:41.335Z","pid":290,"hostname":"resource-history-deployment-6c89769957-mzskj","context":"ResourceEventService","msg":"Received message with resource.ServiceRequest.updated for resource 88068618"}
[resource-history] {"level":30,"time":"2022-11-03T08:36:41.340Z","pid":290,"hostname":"resource-history-deployment-6c89769957-mzskj","context":"ResourceEventService","msg":"Received message with resource.ServiceRequest.updated for resource 88068618"}
[resource-history] {"level":30,"time":"2022-11-03T08:36:41.341Z","pid":290,"hostname":"resource-history-deployment-6c89769957-mzskj","context":"ResourceService","msg":"Registered new resource ServiceRequest/88068618"}
[resource-history] {"level":30,"time":"2022-11-03T08:36:41.343Z","pid":290,"hostname":"resource-history-deployment-6c89769957-mzskj","context":"ResourceService","msg":"**DUPLICATE KEY ERROR MESSAGE**}

I would guess that I need to merge Contexts for the same resources, but not sure how to do this? Or do I need to create a global Request Context that is always open (Sounds stupid) for all messages?

**Comment by B4nan** [comment]
I don't think this is about the EM contexts, this is about what you do inside that handler - you should be running everything in there inside an explicit transaction most probably.

**Comment by rstoermer** [comment]
Thanks for your quick answer! That handler is running for each resource once, so creating a transaction in that handler would still create two transactions, one for each received !same! Resource. So those could still conflict, right?

**Reply by B4nan** [reply]
That's the whole point of transactions, you run two side by side, they collide, one succeeds, one fails, and you just ignore the failure, no? 

I don't have much experience with distributed systems, so can't help much with this problem. Also I have no idea what your service methods are actually doing. It is upsert? How you do it? Note that v5.5 added `em.upsert()`, you might wanna use that - it will run an `insert on conflict update` query.

**Reply by rstoermer** [reply]
Sure :) I thought that maybe I could somehow look up if that entity is already being processed at the moment and merge two RequestContexts for the same Entity. But maybe my thinking is wrong and I need to go over transactions and handle the resulting errors and try again for the failed operation :)  Thanks!

**Comment by rstoermer** [comment]
Okay, getting closer, but don't quite understand how mikro-orm is doing things here. I am doing something like:

```typescript
    console.log('find', id, correlationId, new Date().toISOString());
    const knownResource = await this.resourceRepo.findOne({
      resourceUri,
    });

    let newResource;
    if (!knownResource) {
      newResource = this.createResource(resourceType, afterTrackable);
      await this.resourceRepo.persist(newResource);
    } else {
      knownResource.current = afterTrackable;
    }

    for (const change of someChanges) {
      // Create Changes from compare of before and after trackables
      await this.changesRepo.persist(this.createChange(someArguments));
    }

     console.log(
      'before',
      afterTrackable.id,
      correlationId,
      new Date().toISOString(),
    );
    console.log(
      'flush',
      afterTrackable.id,
      correlationId,
      await this.orm.em.flush(),
    );
    console.log(
      'after',
      afterTrackable.id,
      correlationId,
      new Date().toISOString(),
    );
```

The final flush takes a while to execute, as can be seen in the log.

```
[resource-history] find 88072282 d9c8254ffc55395902e4644d449951e6 2022-11-03T12:31:50.481Z
[resource-history] before 88072282 d9c8254ffc55395902e4644d449951e6 2022-11-03T12:31:50.485Z
[resource-history] flush 88072280 ec408355d8756cb246769ba0d1259c0d undefined
[resource-history] after 88072280 ec408355d8756cb246769ba0d1259c0d 2022-11-03T12:31:50.488Z
[resource-history] {"level":30,"time":"2022-11-03T12:31:50.490Z","pid":104,"hostname":"resource-history-deployment-878b46976-mvqnb","context":"ResourceEventService","msg":"Received message with resource.UnpublishedAppointment.updated for resource UnpublishedAppointment from 2022-11-03T12:31:45.539Z"}
[resource-history] find 88072764 67b025c7ea5b9d3eae97107be7855219 2022-11-03T12:31:50.490Z
[resource-history] before 88072764 67b025c7ea5b9d3eae97107be7855219 2022-11-03T12:31:50.492Z
[resource-history] flush 88072282 d9c8254ffc55395902e4644d449951e6 undefined
[resource-history] after 88072282 d9c8254ffc55395902e4644d449951e6 2022-11-03T12:31:50.512Z
[resource-history] flush 88072764 67b025c7ea5b9d3eae97107be7855219 undefined
[resource-history] after 88072764 67b025c7ea5b9d3eae97107be7855219 2022-11-03T12:31:50.522Z
```

The "after" log always lags behind and takes place after the find for the next resource was done.

**Reply by rstoermer** [reply]
Well, on second thought I must receive multiple messages in parallel and some tasks are already started by Node.js, so this is probably an async problem on my side, rather than mikro-orm waiting for a flush for some reason.

**Reply by B4nan** [reply]
> `await this.resourceRepo.persist(newResource);`

`persist` is sync method, it only makes the EM aware of your entity. flush is what writes to database.

> `await this.changesRepo.persist(this.createChange(someArguments));`

What is inside the `createChange` and what is `someArguments`? That code is very incomplete for me to be able to understand it.

**Reply by rstoermer** [reply]
Oh yeah, right, removed the await :)

createChange is just a Wrapper for the create method of the repository:

```
private createChange(
    changeId: Change['changeId'],
    resourceType: ResourceService['trackableResources'][number],
    resourceId: Trackables['id'],
    changedBy: string,
    op: CompareChange['op'],
    path: CompareChange['path'],
    valueBefore: string,
    value: string,
    timestamp: Date,
  ): Change {
    return this.changesRepo.create({
      changeId,
      resourceUri: `${resourceType}/${resourceId}`,
      resourceType,
      timestamp,
      changedBy: changedBy || null,
      op: op,
      path: path.toString(),
      value,
      valueBefore,
    });
  }
```

**Reply by rstoermer** [reply]
Even if wrapping the above code in a transaction this doesn't help, because even when doing

```
const knownResource = await this.resourceRepo.findOne({
      resourceUri,
    });
```

the resource is not known yet, only after the previous transaction is through a few ms later.

**Reply by B4nan** [reply]
Try `em.upsert` as suggested above

https://mikro-orm.io/docs/entity-manager#upsert

**Reply by rstoermer** [reply]
Yeah this will work! Thank you for your great support and awesome package!! :)
